### PR TITLE
add elementary privilege check for statements

### PR DESCRIPTION
**Pull request by daviszhen**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #3380 
issue #3702

## What this PR does / why we need it:

add elementary privilege check for statements:
1. create/alter/drop account
2. create/alter/drop user
3. create/alter/drop role
4. grant/revoke role
5. use role
6. grant/revoke privilege
7. create/drop database
8. use database
9. create/drop table/view
10. select/update/delete
11. insert/ insert select

The two basic algorithms about the privilege check.
**The algorithm 1**: Decides the privileges of user can satisfy the requirement of the privilege set
Input:
    The userId;
    The default roleId about the userId;
    The privilege Set P;

Steps:
```
step 1: The Set R1 {default role id}
step 2: The Set R2 {the roleid granted to the userid}

Call the algorithm 2.
If the result of the algorithm 2 is true, Then return true;

step 3: !!!NOTE all roleid in setR has been processed by the algorithm 2.
    setVisited is the set of all roleid that has been processed.
    setVisited = setR;
    For {
        For roleA in setR {
            Find the peer roleB in the table mo_role_grant(granted_id,grantee_id) with grantee_id = roleA;
            If roleB is not in setVisited, Then add roleB into setR';
                add roleB into setVisited;
        }

        If setR' is empty, Then return false;
        
        Call the algorithm 2.
        If the result of the algorithm 2 is true, Then return true;
        setR = setR';
        setR' = {};
    }
```

**The algorithm 2**: Decides the privileges of role set can satisfy the requirement of the privilege set.
Input:
    The role Set R;
    The privilege Set P;

Steps:
```
For roleId in R:
    For priv in P:
        Query the privilege tables with (roleId,priv);
        If the roleId has the privilege, it returns true;

Finally, it returns false.
```

### CODE REVIEW

**Review by iamlinjunhong — APPROVED** [review]

File: pkg/frontend/authenticate.go [hunk 1]
```diff
@@ -584,6 +649,105 @@ var (
 	roleIdOfRoleFormat = `select role_id from mo_catalog.mo_role where role_name = "%s";`
 
 	getRoleOfUserFormat = `select r.role_id from  mo_catalog.mo_role r, mo_catalog.mo_user_grant ug where ug.role_id = r.role_id and ug.user_id = %d and r.role_name = "%s";`
+
```
Comment (fengttt) [inline pkg/frontend/authenticate.go:652]: For all these queries, it better runs blazing fast ...

Is there a plan for privilege cache?   It seems to me, that we have some priv info stored in session.  Hope we are ok.

**Review by fengttt — COMMENTED** [review]
Are we computing ALL privilege required to run a command, then check them against all Roles the session has -- looks like via issuing queries?   I cannot see the whole picture.

I am not sure about the performance impact, esp, for small TP queries.    

Can session maintain the privileges that it already acquired.  Next, time, open table etc, just check if the session has the privilege or not ...    Of course,  need to invalidate cache if there is an invoke etc happens.     This can be part of check table version.

**Review by aressu1985 — APPROVED** [review]

**Comment by daviszhen** [comment]
> Are we computing ALL privilege required to run a command, then check them against all Roles the session has -- looks like via issuing queries? I cannot see the whole picture.
> 
> I am not sure about the performance impact, esp, for small TP queries.
> 
> Can session maintain the privileges that it already acquired. Next, time, open table etc, just check if the session has the privilege or not ... Of course, need to invalidate cache if there is an invoke etc happens. This can be part of check table version.

The preliminary picture is that :
1, the statement has the predefined privileges,
2, the role wants to execute the statement checks it has the privileges about the statement or not.
3, the privilege checks are accomplished by querying the tables -- mo_role,mo_user_grant,mo_role_privs.

It influences the performance.

The privilege cache may be a good idea. I do not know how to maintain the consistency between the privilege tables and the cache.

File: pkg/frontend/authenticate.go (hunk 1 above)
Comment (daviszhen) [reply pkg/frontend/authenticate.go:652]: Right now it is just the basic privilege check logic. The privilege cache may be designed next.

**Pull request merged**